src/action.py

```diff
@@ -140,39 +138,46 @@ def _update(
         raise exceptions.ActionError(
             f"internal error, content change for page is None, {action=!r}"
         )
-    if (
-        action.navlink_change.new.link is not None
-        and action.content_change is not None
-        and action.content_change.new is None
-    ):
-        raise exceptions.ActionError(f"internal error, new content for page is None, {action=!r}")
 
-    if (
-        action.content_change is not None
-        and action.content_change.old is not None
-        and action.content_change.new is not None
-    ):
-        _log_content_change(old=action.content_change.old, new=action.content_change.new)
+    if action.content_change is not None:
+        _log_content_change(
+            base=action.content_change.base or action.content_change.server,
+            new=action.content_change.local,
+        )
 
     if (
         not dry_run
         and action.navlink_change.new.link is not None
         and action.content_change is not None
-        and action.content_change.new is not None
-        and action.content_change.new != action.content_change.old
+        and action.content_change.base is not None
+        and action.content_change.local != action.content_change.server
     ):
         try:
-            discourse.update_topic(
-                url=action.navlink_change.new.link, content=action.content_change.new
+            merged_content = content_merge(
+                base=action.content_change.base,
+                theirs=action.content_change.server,
+                ours=action.content_change.local,
             )
+            discourse.update_topic(url=action.navlink_change.new.link, content=merged_content)
             result = types_.ActionResult.SUCCESS
             reason = None
-        except exceptions.DiscourseError as exc:
+        except (exceptions.DiscourseError, exceptions.ContentError) as exc:
             result = types_.ActionResult.FAIL
             reason = str(exc)
     else:
-        result = types_.ActionResult.SKIP if dry_run else types_.ActionResult.SUCCESS
-        reason = DRY_RUN_REASON if dry_run else None
+        if dry_run:
+            result = types_.ActionResult.SKIP
+            reason = DRY_RUN_REASON
+        elif (
+            action.content_change is not None
```

Comment:
It would be good to extract this condition, repeated also in line 151, to a different method for maintainability

Reply:
this condition is actually different, it checks that `action.content_change.base is None` compared to 152 which checks that `action.content_change.base is not None`. Is that what you are referring to?

Reply:
I changed this to make it clearer that action.py executes logic to determine how to execute the action

Reply:
Resolving for now, please let me know if you have further feedback on this